### PR TITLE
Add CUDA, CUDA samples, and CUDA related hooks and lmodrc changes [WIP, do NOT deploy unless sure it works!]

### PR DESCRIPTION
**Pull request by casparvl**
Equivalent of https://github.com/EESSI/software-layer/pull/381, now targetting the new eessi.io domain.

### CODE REVIEW

File: create_lmodrc.py
Comment (ocaisa) [inline create_lmodrc.py:48]: Now that you are shipping the scripts, you could just tell them what script to run

File: create_lmodrc.py
Comment (ocaisa) [inline create_lmodrc.py:64]: Same, you could just ask them to run the script and rely on it being very very helpful

File: create_lmodrc.py
Comment (ocaisa) [inline create_lmodrc.py:69]: It _should_ exist, but what if it doesn't

File: create_lmodrc.py
Comment (ocaisa) [inline create_lmodrc.py:85]: This too could just tell them to rerun the script

**Review by ocaisa — CHANGES_REQUESTED** [review]
There's some really impressive stuff going on here :P

**Comment by ocaisa** [comment]
This doesn't work as is, it requires #410

File: EESSI-install-software.sh
Comment (ocaisa) [inline EESSI-install-software.sh:233]: We need to find a way to allow people to be able to use this script outside of a PR situation

**Comment by ocaisa** [comment]
@casparvl This needs a sync with the target branch

**Comment by casparvl** [comment]
Just want to see what happens if I trigger the bot on this...

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

File: EESSI-install-software.sh
Comment (casparvl) [reply EESSI-install-software.sh:246]: I'm not sure I understand why: what is the use case where end-users or sysadmins would have to re-run this?

File: create_lmodrc.py
Comment (casparvl) [reply create_lmodrc.py:48]: Good point, easy change, I'll do that before we merge.

File: create_lmodrc.py
Comment (casparvl) [reply create_lmodrc.py:69]: I don't know, this is your code XD 
But in all seriousness: you mean we should have an `else` statement here that does _something_ (at least print some error or so?)

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.8.2-system.yml
Comment (boegel) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.8.2-system.yml:8]: This is not a good idea I think, since this isn't a normal CUDA installation, but a stripped down one.
I would prefer to keep track of the CUDA installations we provide in a separate, dedicated, easystack file like `eessi-2023.06-eb-4.8.2-CUDA.yml`, to avoid confusion.

File: create_lmodrc.py
Comment (ocaisa) [reply create_lmodrc.py:73]: No, we don't need the symlinks in a container as an apptainer container already sets LD_LIBRARY_PATH. Hence the second check

File: EESSI-install-software.sh
Comment (ocaisa) [reply EESSI-install-software.sh:246]: Not just for them, also for us, in testing this PR I have to manually edit these lines

Comment (boegel) [reply EESSI-install-software.sh:246]: let's follow up on that in #435

**Comment by boegel** [comment]
CUDA installation in first attempt (job 2696) failed with errors like `/var/log/nvidia/.uninstallManifests/ can't be opened` (which are well hidden between *a lot* of `Log file not open` messages...).

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
--nv was being passed with a space glued to it, which led to "`Error for command "run": unknown flag: --nv `", fixed in 35d6084

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1

File: install_scripts.sh
Comment (ocaisa) [inline install_scripts.sh:54]: I think if we are shipping scripts then the GPU stuff should appear as a subdir in there, I'll try to make a PR tonight

Comment (boegel) [reply install_scripts.sh:54]: I won't let that block this PR, since we have things in a working state to get `CUDA-Samples` installed, so should be in a follow-up PR.

I'm actually very much in favor of not shipping the scripts in the EESSI repository just yet, and to provide instructions in the documentation to clone the `software-layer` repo instead, for now...

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic

File: install_scripts.sh
Comment (casparvl) [reply install_scripts.sh:54]: You mean `/cvmfs/reponame/scripts/gpu_support`? I thought about that as well. I think it is cleaner, but note that the scripts often assume other files to be present in certain relative paths. So you'll have to make sure to adjust these.

I could try to have a look in a minute...

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic

**Comment by boegel** [comment]
Job `2703` failed because of missing `import` statements after refactoring hooks in a79c39f24f747050d84f31e53826beb0a9589de0, fixed with force push in f80f0fc0f70026e4be1fa980e13a991102503737 .

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1

File: install_scripts.sh
Comment (ocaisa) [reply install_scripts.sh:54]: @boegel this PR is already shipping the scripts...

File: create_lmodrc.py
Comment (casparvl) [reply create_lmodrc.py:73]: Discussed with Kenneth, we'll tackle this in a follow-up PR. We need to figure out also for drivers in the `.singularity.d` dir if they are new enough anyway and print informative messages if not. But: that's all further refinement.

File: install_scripts.sh
Comment (ocaisa) [reply install_scripts.sh:54]: For me, I think it's more convenient for the end user to have this self-documenting via Lmod

Comment (boegel) [reply install_scripts.sh:54]: > @boegel this PR is already shipping the scripts...

No, it's not. It's installing the scripts in the build environment, but they are not picked up when creating the tarball for deployment. And even if they would be, the ingestion script running on the Stratum 0 would choke on a tarball that (only) includes stuff in `scripts`.

> For me, I think it's more convenient for the end user to have this self-documenting via Lmod

Why document in two places?
If something goes wrong, the Lmod hook should point to the right page/subsection in the docs (which we need anyway imho).
Also, an Lmod hook that prints a couple of lines and it painful to update isn't the right place to document stuff imho. That's why we created https://eessi.io/docs/gpu as placeholder page.

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.8.2-2023a.yml
Comment (boegel) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.8.2-2023a.yml:38]: remove empty line:
```suggestion
```

**Review by boegel — CHANGES_REQUESTED** [review]

File: install_scripts.sh
Comment (ocaisa) [reply install_scripts.sh:54]: Why make the user go through hoops? Lmod can say:"Run this script"

Comment (ocaisa) [reply install_scripts.sh:54]: The Lmod hook as to error out anyway, why not have it print useful information

File: EESSI-install-software.sh
Comment (ocaisa) [inline EESSI-install-software.sh:193]: @boegel this is intended to install the scripts into CVMFS, you mean that this is not actually working?

File: install_scripts.sh
Comment (ocaisa) [reply install_scripts.sh:54]: And what happens if we change how we do things for a new EESSI version, the Lmod hook would then point to bad jnformation

Comment (boegel) [reply install_scripts.sh:54]: I'm not opposed to providing useful information (or to shipping the scripts in the EESSI repo), I just assume whatever we can print in the Lmod hook won't be *sufficient* for more people.

They'll want to know *what* the script will do exactly, where it will install stuff to, what side effect that has, how to control it, etc.

File: EESSI-install-software.sh
Comment (boegel) [reply EESSI-install-software.sh:193]: It will install scripts in `/cvmfs/software.eessi.io` in the build environment that the bot sets up, but that's not the same as actually deploying those scripts in the EESSI repository on the Stratum 0.

Only stuff that is included in the tarball (which is created with the 🥁 `create_tarball.sh` script in this repo) is actually ingested in the EESSI repo on deployment.
Stuff in `versions/2023.06/scripts/` does not get included in the tarball, currently, so it doesn't get deployed.

File: install_scripts.sh
Comment (boegel) [reply install_scripts.sh:54]: > And what happens if we change how we do things for a new EESSI version, the Lmod hook would then point to bad jnformation

Docs can be updated to have version-specific info, so that's not exactly true.
Lmod hook could be updated to point to `https://eessi.io/docs/gpu/2027.09`, and `https://eessi.io/docs/gpu` could be set up to auto-redirect to `https://eessi.io/docs/gpu/2023.06`.

Comment (ocaisa) [reply install_scripts.sh:54]: You don't think thats a lot of extra effort for little gain?

Personally, I think there are plenty  of people who do not care _how_ it works, just that _it works_.

Comment (ocaisa) [reply install_scripts.sh:54]: You can always add the docs pointers in the scripts if they're actually needed

File: EESSI-install-software.sh
Comment (ocaisa) [reply EESSI-install-software.sh:193]: Ok, but the intention here was to actually do the install, otherwise there is no reason for the installation step

Comment (ocaisa) [reply EESSI-install-software.sh:193]: I'll fix it this evening then

File: install_scripts.sh
Comment (boegel) [reply install_scripts.sh:54]: For some people, just getting it to work will be enough.

Not for system administrators though, they won't go and run a random script just because they're told to.
They may even want to do what needs to be done some other way (through Puppet or whatever).

Comment (ocaisa) [reply install_scripts.sh:54]: Sure but then you are sculpting things for minority use case (sysadmins).

Anyway, we can do both, the Lmod hook can mention the script and the doc, let the users decide

**Comment by boegel** [comment]
Installation of `CUDA-Samples` on `aarch64` (job `2706`) failed with:

```
nvlink error   : Undefined reference to '__cudaCDP2Free' in 'BezierLineCDP.o' (target: sm_87)
nvlink error   : Undefined reference to '__cudaCDP2Malloc' in 'BezierLineCDP.o' (target: sm_87)
nvlink error   : Undefined reference to '__cudaCDP2GetParameterBufferV2' in 'BezierLineCDP.o' (target: sm_87)
nvlink error   : Undefined reference to '__cudaCDP2LaunchDeviceV2' in 'BezierLineCDP.o' (target: sm_87)
make[1]: *** [Makefile:335: cdpBezierTessellation] Error 255
```

Not sure what's causing this, but we'll need to figure it out....

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1

File: EESSI-install-software.sh
Comment (casparvl) [reply EESSI-install-software.sh:193]: Note that https://github.com/EESSI/filesystem-layer/blob/main/scripts/ingest-tarball.sh also may or may not work for this. It _seems_ that it was anticipated that a `scripts` directory would be installed, so probably as long as the top-level dir within the repo is named _exactly_ that, I think it'll work. But it's good to be aware that there could be issues here. In this scenario, I believe it _would_ end up in the tarball, but would still not be ingested by the stratum 0. So in case you encounter that pattern, this is where script which likely should be fixed.

**Comment by ocaisa** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1

**Comment by ocaisa** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1

File: create_lmodrc.py
Comment (boegel) [inline create_lmodrc.py:66]:
```suggestion
            advice = advice .. "to make sure EESSI can find the GPU driver libraries on your host system.\\n"
````

**Review by boegel — CHANGES_REQUESTED** [review]

File: create_tarball.sh
Comment (boegel) [inline create_tarball.sh:46]:
```suggestion
# include scripts that were copied by install_scripts.sh
if [ -d ${eessi_version}/scripts ]; then
    # include scripts we wish to ship along with EESSI,
    find ${eessi_version}/scripts -type f | grep -v '/\.wh\.' >> ${files_list}
fi

```

**Review by boegel — CHANGES_REQUESTED** [review]

File: install_scripts.sh
Comment (boegel) [inline install_scripts.sh:63]: `-u` is used here to only actually copy the script if it was updated, but `cp` will rely on the timestamp of the file for this, which will always be more recent than the files that are already in the repo because the checkout of the repository will be recent.

If we only want to copy if there was a change, we should use `diff` to check if there are any changes, or perhaps use `rsync`?

Not a big issue for now, but with the current approach all scripts that are copied by `install_scripts.sh` will always be included in the tarball being created...

**Comment by casparvl** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1

**Comment by ocaisa** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1

**Comment by ocaisa** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by boegel** [comment]
Upload of tarball for `aarch64/neoverse_v1` by bot failed due to connection error:

<details>

```
[20231221-T14:09:06] WARNING: A crash occurred!
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/usr/lib/python3.6/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3.6/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib64/python3.6/http/client.py", line 1365, in getresponse
    response.begin()
  File "/usr/lib64/python3.6/http/client.py", line 320, in begin
    version, status, reason = self._read_status()
  File "/usr/lib64/python3.6/http/client.py", line 289, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response
```

</details>

I've uploaded `eessi-2023.06-software-linux-aarch64-neoverse_v1-1703166553.tar.gz` manually using the `bot` account...

**Comment by boegel** [comment]
All 8 staging PRs merged, deploy is under way...

**Review by ocaisa — APPROVED** [review]

**Comment by ocaisa** [comment]
Going in, thanks to everyone for all the efforts to get this merged, especially @huebner-m who put a lot of the initial effort in. It's been a journey!

**Pull request merged**
